src/map.cpp

```diff
@@ -755,46 +592,253 @@ const vehicle *map::vehproceed()
         }
     }
 
-    int last_turn_dec = 1;
-    if (veh->last_turn < 0) {
-        veh->last_turn += last_turn_dec;
-        if (veh->last_turn > -last_turn_dec) { veh->last_turn = 0; }
-    } else if (veh->last_turn > 0) {
-        veh->last_turn -= last_turn_dec;
-        if (veh->last_turn < last_turn_dec) { veh->last_turn = 0; }
+    const int last_turn_dec = 1;
+    if( veh.last_turn < 0 ) {
+        veh.last_turn += last_turn_dec;
+        if( veh.last_turn > -last_turn_dec ) {
+            veh.last_turn = 0;
+        }
+    } else if( veh.last_turn > 0 ) {
+        veh.last_turn -= last_turn_dec;
+        if( veh.last_turn < last_turn_dec ) {
+            veh.last_turn = 0;
+        }
     }
 
-    if (can_move) {
+    if( can_move ) {
         // accept new direction
-        if (veh->skidding) {
-            veh->face.init (veh->turn_dir);
-            if(pl_ctrl) {
-                veh->possibly_recover_from_skid();
-            }
+        if( veh.skidding ) {
+            veh.face.init( veh.turn_dir );
         } else {
-            veh->face = mdir;
+            veh.face = facing;
         }
-        veh->move = mdir;
-        if (coll_turn) {
-            veh->skidding = true;
-            veh->turn (coll_turn);
+
+        veh.move = facing;
```

Comment:
This is not the same as `veh->move = mdir;`. This invokes the constructor `tileray(int)`. Same problem a few lines above where you assign it to `veh.face`.

The vehicle now moves in a straight line (directly north, not a bit west or east), even through it's clearly angled sideways (at a degree of 15°, or 30°). In fact, I can only move it along 8 different direction: N, NW, W, SW, and so on. It only moves diagonal if the vehicle faces exactly that direction (45°, 135°, 225°, 315°), otherwise it moves only horizontal or diagonal.

The displayed car rotates just fine with degree of 15° and so on, it just doesn't move that way. Most clearly seen when driving along a straight road with "Auto-shift the view while driving" off and "Draw vehicle facing indicator" on and in a schoolbus.

The tileray class contains x and y values and provides kind of a bresenham line. When `advance` is called, it moves to the next point in that line, but initializing it from an `int`, resets the the line. It's called on each single-step movement and for a movement along the angle 15°, it increments the x value most of the the time, but sometimes the y value:

- x,y = (0, 0)
- advance(), x,y = (1, 0)
- advance(), x,y = (2, 0)
- advance(), x,y = (2, 1)
- advance(), x,y = (3, 1)
- advance(), x,y = (4, 1)
- advance(), x,y = (4, 2)
- advance(), x,y = (5, 2)
and so on.

To do this correctly, it needs to know what point it had before, by converting the `tileray` into a `int` and assigning it, back to a `tileray`, this information gets lost and it will always return the first point of the bresenham line.